datamade/la-metro-dashboard
# Dynamically create scrape DAGs from a config file

## Description

This PR revises our approach to scheduling. Rather than making use of a branch operator, which tightly couples DAG outcomes with the day and time they're run, it moves all scheduling into the cron layer. Because Airflow doesn't natively allow multiple schedule intervals for the same DAG, we achieve this by creating a config object containing each type of scrape and its schedule, then iterating over it in `scrape_factory.py` to dynamically generate the corresponding DAGs. The only exceptions to this pattern are `full_scrape` (which I've refactored into a DAG with a separate task for each scrape, so scraped data is cleared out between each entity type, hopefully addressing #63 and #64) and the ETL DAGs, `hourly_processing` and `refresh_guid`.

Handles #59.

## Testing instructions

- Remove any Airflow volumes you already have: `docker-compose down --volumes`
- Run the Metro app and Airflow instance as described in the README.
- Navigate to the Airflow instance at http://localhost:8080 (logging in with the dummy credentials, `foo` / `bar`) and confirm you see a DAG for every scrape and schedule interval, plus the full scrape and ETL DAGs.

![Screen Shot 2020-12-09 at 9 08 46 AM](https://user-images.githubusercontent.com/12176173/101647285-2a5ac400-39fe-11eb-99cb-980e464d8b61.png)

N.b., I've triggered the smaller scrapes locally to confirm the revised pattern works. After review, I'd like to deploy this to staging and let it run for a couple of days to test that it's working as expected in a deployment environment.
